src/app/3d/qgs3danimationwidget.cpp

```diff
@@ -0,0 +1,278 @@
+    keyframes << kf;
+  }
+  animSettings.setKeyframes( keyframes );
+  return animSettings;
+}
+
+void Qgs3DAnimationWidget::setDefaultAnimation()
+{
+  Qgs3DAnimationSettings animSettings;
+  Qgs3DAnimationSettings::Keyframes kf;
+  Qgs3DAnimationSettings::Keyframe f1, f2;
+  f1.time = 0;
+  f1.point = mCameraController->lookingAtPoint();
+  f1.dist = mCameraController->distance();
+  f1.pitch = mCameraController->pitch();
+  f1.yaw = mCameraController->yaw();
+
+  f2.time = 5;
+  f2.point = f1.point;
+  f2.dist = f1.dist * 2;
+  f2.pitch = f1.pitch;
+  f2.yaw = f1.yaw;
+
+  kf << f1 << f2;
+  animSettings.setKeyframes( kf );
+
+  setAnimation( animSettings );
+}
+
+void Qgs3DAnimationWidget::onPlayPause()
+{
+  if ( mAnimationTimer->isActive() )
+  {
+    mAnimationTimer->stop();
+    cboKeyframe->setEnabled( true );
+  }
+  else
+  {
+    cboKeyframe->setCurrentIndex( 0 );  // unset active keyframe
+    cboKeyframe->setEnabled( false );
+    mAnimationTimer->start();
+  }
+}
+
+void Qgs3DAnimationWidget::onAnimationTimer()
+{
+  float duration = sliderTime->maximum();
+  sliderTime->setValue( sliderTime->value() >= duration ? 0 : sliderTime->value() + 1 );
+}
+
+
+void Qgs3DAnimationWidget::onSliderValueChanged()
+{
+  // make sure we do not have an active keyframe
+  if ( cboKeyframe->currentIndex() != 0 )
+    cboKeyframe->setCurrentIndex( 0 );
+
+  Qgs3DAnimationSettings::Keyframe kf = mAnimationSettings->interpolate( sliderTime->value() / 100. );
+  mCameraController->setLookingAtPoint( kf.point, kf.dist, kf.pitch, kf.yaw );
+}
+
+void Qgs3DAnimationWidget::onCameraChanged()
+{
+  if ( cboKeyframe->currentIndex() <= 0 )
+    return;
+
+  // update keyframe's camera position/rotation
+  int i = cboKeyframe->currentIndex();
+  Qgs3DAnimationSettings::Keyframe kf = cboKeyframe->itemData( i, Qt::UserRole + 1 ).value<Qgs3DAnimationSettings::Keyframe>();
+  kf.point = mCameraController->lookingAtPoint();
+  kf.dist = mCameraController->distance();
+  kf.pitch = mCameraController->pitch();
+  kf.yaw = mCameraController->yaw();
+  cboKeyframe->setItemData( i, QVariant::fromValue<Qgs3DAnimationSettings::Keyframe>( kf ), Qt::UserRole + 1 );
+
+  initializeController( animation() );
+}
+
+void Qgs3DAnimationWidget::onKeyframeChanged()
+{
+  if ( cboKeyframe->currentIndex() <= 0 )
+    return;
+
+  // jump to the camera view of the keyframe
+  Qgs3DAnimationSettings::Keyframe kf = cboKeyframe->itemData( cboKeyframe->currentIndex(), Qt::UserRole + 1 ).value<Qgs3DAnimationSettings::Keyframe>();
+
+  whileBlocking( sliderTime )->setValue( kf.time * 100 );
+  mCameraController->setLookingAtPoint( kf.point, kf.dist, kf.pitch, kf.yaw );
+}
+
+
+void Qgs3DAnimationWidget::onAddKeyframe()
+{
+  bool ok;
+  double t = QInputDialog::getDouble( this, tr( "Add keyframe" ), tr( "Keyframe time [seconds]:" ), sliderTime->value() / 100., 0, 9999, 2, &ok );
+  if ( !ok )
+    return;
+
+  // figure out position of this keyframe
+  int index = 0;
```

Comment:
Is it worth moving this to an "addKeyFrame" method in Qgs3DAnimationSettings?

Reply:
Not sure about that - the code does not alter the animation settings directly (it just inserts into a combo box which is used to recreate the animation settings) + there is GUI code inside...

(but yes in general it would be useful to have add+remove+update methods for keyframes in addition to the setKeyframes() call)